File: ui/tract-renderer/src/main/java/org/cru/godtools/tract/viewmodel/ButtonViewHolder.java
package org.cru.godtools.tract.viewmodel;

import android.content.Context;
import android.content.res.ColorStateList;
import android.net.Uri;
import android.view.ViewGroup;
import android.widget.TextView;

import org.cru.godtools.base.tool.analytics.model.ExitLinkActionEvent;
import org.cru.godtools.base.tool.model.view.TextViewUtils;
import org.cru.godtools.base.ui.util.WebUrlLauncher;
import org.cru.godtools.tract.R;
import org.cru.godtools.tract.R2;
import org.cru.godtools.xml.model.AnalyticsEvent.Trigger;
import org.cru.godtools.xml.model.Button;
import org.cru.godtools.xml.model.Text;
import org.greenrobot.eventbus.EventBus;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import androidx.core.view.ViewCompat;
import butterknife.BindView;
import butterknife.OnClick;

@UiThread
final class ButtonViewHolder extends BaseViewHolder<Button> {
    @BindView(R2.id.button)
    TextView mButton;

    ButtonViewHolder(@NonNull final ViewGroup parent, @Nullable final BaseViewHolder parentViewHolder) {
        super(Button.class, parent, R.layout.tract_content_button, parentViewHolder);
    }

    // region Lifecycle Events

    @Override
    void onBind() {
        super.onBind();
        final Text text = mModel != null ? mModel.getText() : null;
        TextViewUtils.bind(text, mButton);
        ViewCompat.setBackgroundTintList(mButton, ColorStateList.valueOf(Button.getButtonColor(mModel)));
    }

    // endregion Lifecycle Events

    @OnClick(R2.id.button)
    void click() {
        if (mModel != null) {
            triggerAnalyticsEvents(mModel.getAnalyticsEvents(), Trigger.SELECTED, Trigger.DEFAULT);

            switch (mModel.getType()) {
                case URL:
                    final Uri url = mModel.getUrl();
                    if (url != null) {
                        final Context context = mRoot.getContext();
                        EventBus.getDefault().post(new ExitLinkActionEvent(url));
                        WebUrlLauncher.openUrl(context, url);
                    }
                    break;
                case EVENT:
                    sendEvents(mModel.getEvents());
                    break;
            }
        }
    }
}

File: ui/tract-renderer/src/main/java/org/cru/godtools/tract/viewmodel/ButtonViewHolder.kt
package org.cru.godtools.tract.viewmodel

import android.view.ViewGroup
import androidx.annotation.UiThread
import org.cru.godtools.analytics.model.ExitLinkActionEvent
import org.cru.godtools.base.ui.util.openUrl
import org.cru.godtools.tract.R
import org.cru.godtools.tract.databinding.TractContentButtonBinding
import org.cru.godtools.xml.model.AnalyticsEvent.Trigger
import org.cru.godtools.xml.model.Button
import org.greenrobot.eventbus.EventBus

@UiThread
internal class ButtonViewHolder(parent: ViewGroup, parentViewHolder: BaseViewHolder<*>?) :
    BaseViewHolder<Button>(Button::class.java, parent, R.layout.tract_content_button, parentViewHolder) {
    private val binding = TractContentButtonBinding.bind(mRoot).also { it.holder = this }

    public override fun onBind() {
        super.onBind()
        binding.model = mModel
    }

    fun click() {
        val model = mModel ?: return

        triggerAnalyticsEvents(model.analyticsEvents, Trigger.SELECTED, Trigger.DEFAULT)
        when (model.type) {
            Button.Type.URL -> model.url?.let { url ->
                EventBus.getDefault().post(ExitLinkActionEvent(url))
                mRoot.context.openUrl(url)
            }
            Button.Type.EVENT -> sendEvents(model.events)
        }
    }
}
